integrations/git_annex_test.go

```diff
@@ -0,0 +1,940 @@
+
+TODO: this could be replaced with a fixture repo;
+see integrations/gitea-repositories-meta/ and models/fixtures/repository.yml.
+
+However we reuse this template for -different- repos.
+*/
+func doInitAnnexRepository(t *testing.T, repoPath string) {
+	// set up what files should be annexed
+	// in this case, all *.bin  files will be annexed
+	// without this, git-annex's default config annexes every file larger than some number of megabytes
+	f, err := os.Create(path.Join(repoPath, ".gitattributes"))
+	require.NoError(t, err)
+	f.WriteString("*                   annex.largefiles=nothing")
+	f.WriteString("*.bin  filter=annex annex.largefiles=anything")
+	f.Close()
+	require.NoError(t, git.AddChanges(repoPath, false, "."))
+	require.NoError(t, git.CommitChanges(repoPath, git.CommitChangesOptions{Message: "Configure git-annex settings"}))
+
+	// 'git annex init'
+	// 'gitea-annex-test' is there to avoid the nuisance comment getting stored.
+	require.NoError(t, git.NewCommand(git.DefaultContext, "annex", "init", "gitea-annex-test").Run(&git.RunOpts{Dir: repoPath}))
+
+	// add a file to the annex
+	require.NoError(t, generateRandomFile(1024*1024/4, path.Join(repoPath, "large.bin")))
+	require.NoError(t, git.AddChanges(repoPath, false, "."))
+	require.NoError(t, git.CommitChanges(repoPath, git.CommitChangesOptions{Message: "Annex a file"}))
+}
+
+/* given a git repo and a path to an annexed file in it (assumed to be committed to its HEAD),
+   find the path in .git/annex/objects/ that contains its actual contents. */
+func annexObjectPath(repoPath string, file string) (string, error) {
+
+	// find the path inside *.git/annex/objects of a given file
+	// i.e. figure out its two-level hash prefix: https://git-annex.branchable.com/internals/hashing/
+	// ASSUMES the target file is checked into HEAD
+
+	var bare bool // whether the repo is bare or not; this changes what the hashing algorithm is, due to backwards compatibility
+
+	bareStr, _, err := git.NewCommand(git.DefaultContext, "config", "core.bare").RunStdString(&git.RunOpts{Dir: repoPath})
+	if err != nil {
+		return "", err
+	}
+
+	if bareStr == "true\n" {
+		bare = true
+	} else if bareStr == "false\n" {
+		bare = false
+	} else {
+		return "", errors.New(fmt.Sprintf("Could not determine if %s is a bare repo or not; git config core.bare = <%s>", repoPath, bareStr))
+	}
+
+	// given a repo and a file in it
+	// TODO: handle other branches, e.g. non-HEAD branches etc
+	annexKey, _, err := git.NewCommand(git.DefaultContext, "show", "HEAD:"+file).RunStdString(&git.RunOpts{Dir: repoPath})
+	if err != nil {
+		return "", err
+	}
+
+	annexKey = strings.TrimSpace(annexKey)
+	if !strings.HasPrefix(annexKey, "/annex/objects/") {
+		return "", errors.New(fmt.Sprintf("%s/%s does not appear to be annexed .", repoPath, file))
+	}
+	annexKey = strings.TrimPrefix(annexKey, "/annex/objects/")
+
+	var keyformat string
+	if bare {
+		keyformat = "hashdirlower"
+	} else {
+		keyformat = "hashdirmixed"
+	}
```

Comment:
Nit: since this is the only use of the `bare` variable, why not put `keyformat = "hashdirlower"` instead of `bare = true` above, and same for `bare = false`?

Reply:
In this one instance, I think that would be a lot less clear. The relationship is due to obscure legacy compatibility concerns: https://git-annex.branchable.com/internals/hashing/. I'm hoping that someone reading this, then that link, then this again will be able to tell why this has to be here and I think inlining it would obscure the reasoning. It would for me at least.